### PR TITLE
CoreOS Model

### PR DESCRIPTION
**Pull request by hvolkmer**
To install a CoreOS node, set up hanlon with the following commands (tested in VirtualBox, hence the tag. YMMV):

```
hanlon image add -t mk -p /vagrant/isos/hnl_mk_dev-image.2.0.1.iso
hanlon image add -t os -p /vagrant/isos/coreos_557.iso -v 557
IMAGE_UUID=$(hanlon image | grep coreos | awk '{print $1}')
hanlon model add --template coreos_stable --label install_coreos --image-uuid $IMAGE_UUID -o /vagrant/hanlon-config/model.yaml
MODEL_UUID=$(hanlon model | grep coreos | awk '{print $5}')
hanlon policy add --template linux_deploy --label coreos --model-uuid $MODEL_UUID  --broker-uuid=none --tags OracleCorporation -e true
```

For `model.yaml` use this:

```
hostname_prefix: "coreosnode"
domainname: localdomain
cloud_config:
  coreos:
    units:
      - name: etcd.service
        command: start
      - name: fleet.service
        command: start
  ssh_authorized_keys:
    - ssh-rsa AAAAB3NzaC1y....yourkey
install_disk: '/dev/sda'
```

The cloudconfig part is based on [CoreOS Cloud Config](https://coreos.com/docs/cluster-management/setup/cloudinit-cloud-config/)

### CODE REVIEW

**Comment by tjmcs** [comment]
Thanks for the PR, @hvolkmer; this actually resolves an issue that was raised by @bunchc last September (New Model Request - CoreOS #177). I think @mtnbikenc is busy reviewing this PR as we speak; hopefully we'll have his feedback fairly quickly. From my brief review of this PR, however, I think it looks good :)

There is one issue we need to resolve from your side, however. Before we can merge this PR, we need to get a CLA signature from you. The process is completely electronic, and the form is available [here](http://opensource.csc.com/sysworkflow/en/neoclassic/251810809537eb36f73ac23031915862/Signing_CLA_Welcome_Page.php). If you could get that CLA signed and let us know here in the comments that you have signed it, we'd love to get this PR merged (and get support for CoreOS into Hanlon).

**Comment by hvolkmer** [comment]
Hi @tjmcs, I tried to go through the CLA process. It ended with an error:

```
An error occurred while the application was being processed.
Error code: 14
Error message: Task invalid or the user is not assigned to the task 
```

**Comment by tjmcs** [comment]
my apologies...let me check with the folks responsible for that system and see if we can get the error resolved...in the meantime, I'll also check and see if we can just accept your signature as is (since you obviously tried to sign) ;)

**Comment by mtnbikenc** [comment]
Looks like you have some unmatched curly braces and I don't think the 'cloud_config' section of the optional yaml file is working as you expect.  At least it is not no my end.  I'll make a couple of code comments.

File: core/model/coreos.rb [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            :validation  => '^[a-zA-Z0-9][a-zA-Z0-9\-]*$',
+            :required    => true,
+            :description => "node hostname prefix (will append node number)"
+          },
+          "@domainname" => {
+            :default     => "localdomain",
+            :example     => "example.com",
+            :validation  => '^[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9](\.[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])*$',
+            :required    => true,
+            :description => "local domain name (will be used in /etc/hosts file)",
```
Comment (mtnbikenc) [inline core/model/coreos.rb]: Missing closing curly brace for domainname.

File: core/model/coreos.rb [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+          "@install_disk" => { 
+            :default      => "/dev/sda",
+            :example      => "/dev/sda",
+            :validation   => '',
+            :required     => false,
+            :description  => "The Core OS target disk" },
+          }
```
Comment (mtnbikenc) [inline core/model/coreos.rb]: Extra curly brace here.

**Comment by hvolkmer** [comment]
Ah. Yes... I moved the install disk to req_metadata from optional and messed up the curly braces.

**Comment by hvolkmer** [comment]
@mtnbikenc What do you mean by " I don't think the 'cloud_config' section of the optional yaml file is working as you expect"? I used the file as shown above and the ssh key was installed by core os and the services where running...

**Comment by mtnbikenc** [comment]
@hvolkmer The 'cloud_config' section of the optional metadata yaml file was not getting imported into the model when I was adding it before.  With the corrections of the curly braces that has been resolved.

**Comment by mtnbikenc** [comment]
I was able to deploy a CoreOS node and log in using ssh keys.  Looks good to merge.

:+1:

**Comment by tjmcs** [comment]
@hvolkmer, looks like the code is working at this end, but I had one last question for you...is there any particular reason why the install disk is in the required metadata parameters instead of the optional metadata parameters? It just seems to me that this would more closely resemble the other Linux models if the user didn't have to specify the install disk (but could specify it if they wanted it to be something other than the first drive)...thoughts?

**Comment by hvolkmer** [comment]
I haven't looked at the other models in detail.... We could also define /dev/sda as default and override via optional parameter. If this makes everything more consistent, why not.

Let me know if I should change it and I will push another commit

**Comment by tjmcs** [comment]
I think it'd be cleaner...setup a default at the start of the `core/model/coreos.rb` file...say somewhere around line 36, stating that `@install_disk = '/dev/sda'` (you might want to add a default for the `@cloud_config` class variable there too), then move lines 53-59 (from the current version of this file) down into the `@opt_metadata_hash` section.

The `@opt_metadata_hash` section is something we added relatively recently (to support optional parameters you might want to include in a model but that aren't necessarily required for all instances of that model type). We use it when defining VSAN clusters in ESXi installs, for example, and have even discussed using it to add the ability to define custom partitioning schemes in the kickstart files of RHEL/CentOS models (or the equivalents in other Linux models).

One last change I think you might be able to make. In lines 261-267 of that same `core/model/coreos.rb` file you (re)define the `api_svc_uri` and `image_svc_uri` methods. In our recently merged PR that added support for the new Windows model, we actually refactored things a bit to support one central location where these methods are defined; lines 223-229 of the `core/model/base.rb` file now look like this:
```ruby
      def api_svc_uri
        "http://#{config.hanlon_server}:#{config.api_port}#{config.websvc_root}"
      end

      def image_svc_uri
        "#{api_svc_uri}/image/#{@image_prefix}"
      end
```
I think you'll find that with these two methods defined centrally, you don't need to redefine them in the `core/model/coreos.rb` file.

**Comment by hvolkmer** [comment]
Ok. Will look into these tomorrow, update the code and push a new commit

**Comment by tjmcs** [comment]
any updates on this, @hvolkmer?

**Comment by hvolkmer** [comment]
Hi @tjmcs, yes. See [last commit](https://github.com/hvolkmer/Hanlon/commit/d2f623fd738c432e0329cc57a0c48a08c5399884)

**Comment by tjmcs** [comment]
Sorry...missed that commit...getting older, I guess ;)  With those changes in place, I think I'm good to go; :+1: for a merge from me, unless anyone objects I'll get this merged shortly

**Comment by mtnbikenc** [comment]
Successfully deployed a coreos node and verified access with ssh keys.
:+1: for merge

**Comment by tjmcs** [comment]
thx, @hvolkmer (and @mtnbikenc); will merge shortly

**Comment by hvolkmer** [comment]
Thanks!

I was just checking the [docker image build](https://registry.hub.docker.com/u/cscdock/hanlon/builds_history/81705/) which seems to be broken. It looks like it related to [PR 330](https://github.com/csc/Hanlon/issues/330). No idea where to debug this further though. 

I'm using hanlon in a docker image that's why I'm interested in the docker build succeeding :)

**Comment by mtnbikenc** [comment]
Thanks for the tip on the docker build.  It is failing on the ruby 2.2.0 package as it is not available. :(  I'll get that sorted out.

**Pull request merged**